rl4co/models/zoo/deepaco/antsystem.py

```diff
@@ -0,0 +1,245 @@
+from functools import lru_cache
+from typing import Optional, Tuple
+
+import torch
+
+from tensordict import TensorDict
+from torch import Tensor
+
+from rl4co.envs import RL4COEnvBase
+from rl4co.models.nn.dec_strategies import Sampling
+from rl4co.models.zoo.common.nonautoregressive.decoder import (
+    NonAutoregressiveDecoder as NARDecoder,
+)
+
+
+class AntSystem:
+    """Implements the Ant System algorithm: https://doi.org/10.1109/3477.484436.
+
+    Args:
+        log_heuristic: Logarithm of the heuristic matrix.
+        n_ants: Number of ants to be used in the algorithm. Defaults to 20.
+        alpha: Importance of pheromone in the decision-making process. Defaults to 1.0.
+        beta: Importance of heuristic information in the decision-making process. Defaults to 1.0.
+        decay: Rate at which pheromone evaporates. Should be between 0 and 1. Defaults to 0.95.
+        pheromone: Initial pheromone matrix. Defaults to `torch.ones_like(log_heuristic)`.
+        require_logp: Whether to require the log probability of actions. Defaults to False.
+    """
+
+    def __init__(
+        self,
+        log_heuristic: Tensor,
+        n_ants: int = 20,
+        alpha: float = 1.0,
+        beta: float = 1.0,
+        decay: float = 0.95,
+        pheromone: Optional[Tensor] = None,
+        require_logp: bool = False,
+    ):
+        self.batch_size = log_heuristic.shape[0]
```

Comment:
[Very Minor, No need to fix now] theoretically we support multi `batch_size` dimensions. But this rarely happens  and we have to touch a few other functions to fully support that. I think it's fine to keep this.